src/lib.rs

```diff
@@ -314,59 +324,75 @@ where
-                for (k, v) in data.iter() {
-                    if v.expire.load(atomic::Ordering::Relaxed) {
-                        delete_keys.push(k.clone());
-                    } else {
-                        v.expire.store(true, atomic::Ordering::Relaxed);
-                    }
+                {
+                    let data = self.inner.data.read();
+                    keys = data.keys().cloned().collect();
+                    drop(data);
                 }
 
-                for k in delete_keys {
-                    let ety = data.remove(&k).unwrap();
-                    self.send_delete(k, ety.val);
+                // after that, fetch all data using the keys
+                let mut futures = FuturesOrdered::new();
+                for k in &keys {
+                    let fut = self.inner.options.fetcher.fetch(k.clone());
+                    futures.push_back(fut);
                 }
 
-                // then, get all keys
-                keys = data.keys().cloned().collect();
-                drop(data);
-            }
-
-            // after that, fetch all data using the keys
-            let mut futures = FuturesOrdered::new();
-            for k in &keys {
-                let fut = self.inner.options.fetcher.fetch(k.clone());
-                futures.push_back(fut);
-            }
+                // and save them into a new vec
+                let mut new_data = Vec::with_capacity(futures.len());
+                assert!(futures.len() == keys.len());
+                let mut key_index = 0;
+                while let Some(res) = futures.next().await {
+                    let key = unsafe { keys.get_unchecked(key_index) }.clone();
+                    key_index += 1;
+                    match res {
+                        Ok(v) => new_data.push(Some(v)),
+                        Err(e) => {
+                            self.send_err(key, e).await;
+                            new_data.push(None);
+                        }
+                    }
+                }
 
-            // and save them into a new vec
-            let mut new_data = Vec::with_capacity(futures.len());
-            assert!(futures.len() == keys.len());
-            let mut key_index = 0;
-            while let Some(res) = futures.next().await {
-                let key = unsafe { keys.get_unchecked(key_index) }.clone();
-                key_index += 1;
-                match res {
-                    Ok(v) => new_data.push(Some(v)),
-                    Err(e) => {
-                        self.send_err(key, e).await;
-                        new_data.push(None);
+                // finally, replace the old data with the new one
+                let mut data = self.inner.data.write();
+                for (k, v) in keys.into_iter().zip(new_data.into_iter()) {
+                    if let Some(v) = v {
+                        data.get_mut(&k).unwrap().val = v;
```

Comment:
May panic if the `expire_cron` deleted the key in that period.
Need to check if the entry is `None` here, if it is `None` just skip that key.